File: android/ai-app-bridge-android/src/main/kotlin/io/github/mobileaidev/aiappbridge/android/capture/FactStoreReceiptPort.kt
package io.github.mobileaidev.aiappbridge.android.capture

import io.github.mobileaidev.aiappbridge.android.SegmentedFactRecordEnqueueResult
import io.github.mobileaidev.aiappbridge.android.SegmentedFactStore
import io.github.mobileaidev.aiappbridge.android.SegmentedFactStoreCursor
import io.github.mobileaidev.aiappbridge.android.SegmentedFactStoreDurability
import io.github.mobileaidev.aiappbridge.android.SegmentedFactStoreReadResult
import io.github.mobileaidev.aiappbridge.android.SegmentedFactStoreRecord
import io.github.mobileaidev.aiappbridge.android.SegmentedFactStoreResultCode
import org.json.JSONObject
import java.io.File
import java.io.RandomAccessFile
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

data class FactReceipt(
    val status: String,
    val accepted: Boolean,
    val committed: Boolean,
    val storeGeneration: Long,
    val mobileFactId: String?,
    val globalSequence: Long?,
)

data class FactCursor(
    val storeGeneration: Long,
    val afterSequence: Long = 0,
)

data class FactPageItem(
    val mobileFactId: String,
    val payload: ByteArray,
    val partitionId: Int,
    val globalSequence: Long,
    val committed: Boolean,
)

data class FactPage(
    val items: List<FactPageItem>,
    val storeGeneration: Long,
    val generationMismatch: Boolean = false,
)

data class DrainResult(
    val ok: Boolean,
    val committed: Int,
)

data class ThroughWatermark(
    val throughSequence: Long,
    val storeGeneration: Long,
)

internal class FactStoreReceiptPort(
    private val store: SegmentedFactStore,
    private val directory: File,
) {
    private val lock = Any()
    private val pending = LinkedHashMap<String, PendingFact>()
    private val aliases = LinkedHashMap<String, String>()
    private val dropped = LinkedHashSet<String>()
    private val index = LinkedHashMap<String, IndexEntry>()
    private var generation = 1L
    private var sequence = 0L
    private var generationFloor = 0L
    private var storeHighWater = 0L
    private val sidecar: File = File(directory, SIDECAR_NAME)

    init {
        directory.mkdirs()
        loadSidecar()
        hydrateFromStore()
        storeHighWater = index.values.maxOfOrNull { it.globalSequence } ?: 0L
    }

    fun appendWithReceipt(
        payload: ByteArray,
        partitionId: Int = 0,
        durability: String = "async",
    ): FactReceipt {
        val enqueue = store.recordForReceipt(payload, partitionId, durabilityOf(durability))
        if (enqueue != SegmentedFactRecordEnqueueResult.ACCEPTED) {
            return receiptOf(enqueue)
        }
        synchronized(lock) {
            sequence += 1
            val mobileFactId = "mf1:$generation:$sequence:${hashPrefix(payload)}"
            pending[mobileFactId] = PendingFact(mobileFactId, payload.copyOf(), partitionId, sequence)
            return FactReceipt(
                status = "accepted",
                accepted = true,
                committed = false,
                storeGeneration = generation,
                mobileFactId = mobileFactId,
                globalSequence = null,
            )
        }
    }

    fun commitWait(mobileFactId: String, timeoutMs: Long): FactReceipt {
        if (!drainWriter(timeoutMs)) {
            synchronized(lock) {
                return receiptAfterWait(mobileFactId, drained = false)
            }
        }
        promoteCommitted(markUnmatchedDropped = true)
        synchronized(lock) {
            return receiptAfterWait(mobileFactId, drained = true)
        }
    }

    fun readPage(cursor: FactCursor, limit: Int): FactPage {
        synchronized(lock) {
            if (cursor.storeGeneration != generation) {
                return FactPage(emptyList(), generation, generationMismatch = true)
            }
            val items = ArrayList<FactPageItem>(limit)
            for (entry in index.values) {
                if (entry.globalSequence <= cursor.afterSequence) continue
                if (items.size == limit) break
                items.add(entry.toItem())
            }
            for (held in pending.values) {
                if (items.size == limit) break
                items.add(held.toItem())
            }
            return FactPage(items, generation)
        }
    }

    fun flushDrain(timeoutMs: Long): DrainResult {
        if (!awaitFlush(timeoutMs)) return DrainResult(ok = false, committed = 0)
        val before: Int
        synchronized(lock) { before = index.size }
        promoteCommitted(markUnmatchedDropped = true)
        val after: Int
        synchronized(lock) { after = index.size }
        return DrainResult(ok = true, committed = after - before)
    }

    fun clear(): Long {
        val floor = readAllStoreRecords().maxOfOrNull { it.sequence } ?: 0L
        store.takeReceiptOutcomes()
        synchronized(lock) {
            generation += 1
            sequence = 0
            generationFloor = maxOf(generationFloor, floor)
            storeHighWater = maxOf(storeHighWater, generationFloor)
            pending.clear()
            aliases.clear()
            dropped.clear()
            index.clear()
            appendSidecarLine(
                JSONObject()
                    .put("op", "clear")
                    .put("generation", generation)
                    .put("throughSequence", generationFloor)
                    .toString() + "\n",
            )
            return generation
        }
    }

    fun storeGeneration(): Long = synchronized(lock) { generation }

    fun throughWatermark(): ThroughWatermark = synchronized(lock) {
        ThroughWatermark(
            throughSequence = index.values.maxOfOrNull { it.globalSequence } ?: 0L,
            storeGeneration = generation,
        )
    }

    private fun receiptAfterWait(mobileFactId: String, drained: Boolean): FactReceipt {
        if (dropped.contains(mobileFactId)) {
            return dropped(mobileFactId, generation)
        }
        pending[mobileFactId]?.let { return it.toReceipt(generation) }
        aliases[mobileFactId]?.let { aliased -> index[aliased]?.let { return it.toReceipt(generation) } }
        index[mobileFactId]?.let { return it.toReceipt(generation) }
        if (drained) {
            return dropped(mobileFactId, generation)
        }
        return dropped(mobileFactId, generation)
    }

    private fun promoteCommitted(markUnmatchedDropped: Boolean) {
        val records = readAllStoreRecords()
        val outcomes = store.takeReceiptOutcomes()
        synchronized(lock) {
            val taken = index.values.map { it.globalSequence }.toMutableSet()
            val fresh = records
                .filter { record -> record.sequence > generationFloor && record.sequence !in taken }
                .sortedBy { it.sequence }
            val successCount = outcomes.count { it }
            val newlyWritten = if (successCount == 0) emptyList() else fresh.takeLast(successCount)
            val preexisting = if (successCount == 0) fresh else fresh.dropLast(successCount)
            val remaining = LinkedHashMap(pending)
            val promoted = ArrayList<IndexEntry>()
            for (record in preexisting) {
                promoted.add(commitStoreRecord(record, null))
            }
            var writtenIndex = 0
            var outcomeIndex = 0
            for (id in pending.keys.toList()) {
                if (!remaining.containsKey(id)) continue
                if (outcomeIndex >= outcomes.size) break
                val success = outcomes[outcomeIndex]
                outcomeIndex += 1
                remaining.remove(id)
                if (!success) {
                    if (markUnmatchedDropped) dropped.add(id)
                    else remaining[id] = pending[id]!!
                    continue
                }
                if (writtenIndex >= newlyWritten.size) {
                    if (markUnmatchedDropped) dropped.add(id)
                    continue
                }
                val record = newlyWritten[writtenIndex]
                writtenIndex += 1
                promoted.add(commitStoreRecord(record, id))
            }
            while (writtenIndex < newlyWritten.size) {
                promoted.add(commitStoreRecord(newlyWritten[writtenIndex], null))
                writtenIndex += 1
            }
            if (markUnmatchedDropped) {
                for (id in remaining.keys) {
                    dropped.add(id)
                }
                remaining.clear()
            }
            pending.clear()
            pending.putAll(remaining)
            storeHighWater = maxOf(
                storeHighWater,
                promoted.maxOfOrNull { it.globalSequence } ?: 0L,
            )
            for (entry in promoted) {
                appendSidecarLine(entry.toJson().toString() + "\n")
            }
        }
    }

    private fun commitStoreRecord(record: SegmentedFactStoreRecord, acceptId: String?): IndexEntry {
        val hash = hashPrefix(record.payload)
        val committedId = "mf1:$generation:${record.sequence}:$hash"
        if (acceptId != null && acceptId != committedId) {
            aliases[acceptId] = committedId
        }
        val entry = IndexEntry(
            committedId,
            record.payload.copyOf(),
            record.partitionId,
            record.sequence,
            generation,
        )
        index[committedId] = entry
        return entry
    }

    private fun appendSidecarLine(line: String) {
        if (sidecar.isFile && sidecar.length() > 0L) {
            RandomAccessFile(sidecar, "rw").use { raf ->
                raf.seek(raf.length() - 1)
                if (raf.read() != '\n'.code) {
                    raf.write('\n'.code)
                }
            }
        }
        sidecar.appendText(line)
    }

    private fun readAllStoreRecords(): List<SegmentedFactStoreRecord> {
        val records = ArrayList<SegmentedFactStoreRecord>()
        var cursor = SegmentedFactStoreCursor()
        while (true) {
            val page = awaitRead(cursor) ?: break
            val record = page.record
            if (page.isEnd || record == null) break
            records.add(record)
            cursor = page.cursor
        }
        return records
    }

    private fun drainWriter(timeoutMs: Long): Boolean {
        val latch = CountDownLatch(1)
        store.status { latch.countDown() }
        return latch.await(timeoutMs, TimeUnit.MILLISECONDS)
    }

    private fun awaitFlush(timeoutMs: Long): Boolean {
        val latch = CountDownLatch(1)
        var ok = false
        store.flush {
            ok = it.isSuccess
            latch.countDown()
        }
        if (!latch.await(timeoutMs, TimeUnit.MILLISECONDS)) return false
        return ok
    }

    private fun awaitRead(cursor: SegmentedFactStoreCursor): SegmentedFactStoreReadResult? {
        var capacity = 64 * 1024
        repeat(2) {
            val page = awaitReadOnce(cursor, capacity) ?: return null
            if (page.operation.code != SegmentedFactStoreResultCode.BUFFER_TOO_SMALL) return page
            if (page.requiredCapacity <= capacity) return page
            capacity = page.requiredCapacity
        }
        return awaitReadOnce(cursor, capacity)
    }

    private fun awaitReadOnce(
        cursor: SegmentedFactStoreCursor,
        bufferCapacity: Int,
    ): SegmentedFactStoreReadResult? {
        val latch = CountDownLatch(1)
        var value: SegmentedFactStoreReadResult? = null
        store.read(cursor, bufferCapacity) {
            value = it
            latch.countDown()
        }
        if (!latch.await(5, TimeUnit.SECONDS)) return null
        return value
    }

    private fun loadSidecar() {
        if (!sidecar.isFile) return
        sidecar.useLines { lines ->
            for (line in lines) {
                if (line.isEmpty()) continue
                val row = try {
                    JSONObject(line)
                } catch (_: Exception) {
                    continue
                }
                if (row.optString("op") == "clear") {
                    generation = row.getLong("generation")
                    sequence = 0
                    generationFloor = maxOf(generationFloor, row.optLong("throughSequence", 0L))
                    index.clear()
                    continue
                }
                val entry = IndexEntry(
                    mobileFactId = row.getString("mobileFactId"),
                    payload = ByteArray(0),
                    partitionId = row.getInt("partitionId"),
                    globalSequence = row.getLong("globalSequence"),
                    generation = row.getLong("generation"),
                )
                if (entry.generation == generation) {
                    index[entry.mobileFactId] = entry
                    sequence = maxOf(sequence, entry.localSequence())
                }
            }
        }
    }

    private fun hydrateFromStore() {
        if (index.isEmpty()) return
        val records = readAllStoreRecords()
        synchronized(lock) {
            val bySequence = records.associateBy { it.sequence }
            val hydrated = LinkedHashMap<String, IndexEntry>()
            for ((id, entry) in index) {
                val record = bySequence[entry.globalSequence] ?: continue
                if (hashPrefix(record.payload) != hashFromId(id)) continue
                hydrated[id] = IndexEntry(
                    mobileFactId = entry.mobileFactId,
                    payload = record.payload.copyOf(),
                    partitionId = record.partitionId,
                    globalSequence = record.sequence,
                    generation = entry.generation,
                )
            }
            index.clear()
            index.putAll(hydrated)
        }
    }

    private fun receiptOf(enqueue: SegmentedFactRecordEnqueueResult): FactReceipt {
        val status = when (enqueue) {
            SegmentedFactRecordEnqueueResult.QUEUE_FULL -> "queue-full"
            SegmentedFactRecordEnqueueResult.CLOSED -> "closed"
            SegmentedFactRecordEnqueueResult.DISABLED -> "disabled"
            SegmentedFactRecordEnqueueResult.PAYLOAD_TOO_LARGE -> "payload-too-large"
            SegmentedFactRecordEnqueueResult.ACCEPTED -> "accepted"
        }
        return FactReceipt(
            status = status,
            accepted = false,
            committed = false,
            storeGeneration = synchronized(lock) { generation },
            mobileFactId = null,
            globalSequence = null,
        )
    }

    private fun dropped(mobileFactId: String, storeGeneration: Long): FactReceipt = FactReceipt(
        status = "dropped",
        accepted = false,
        committed = false,
        storeGeneration = storeGeneration,
        mobileFactId = mobileFactId,
        globalSequence = null,
    )

    private fun durabilityOf(value: String): SegmentedFactStoreDurability {
        if (value == "sync") return SegmentedFactStoreDurability.SYNC
        if (value == "async") return SegmentedFactStoreDurability.MEMORY
        throw IllegalArgumentException(value)
    }

    private class PendingFact(
        val mobileFactId: String,
        val payload: ByteArray,
        val partitionId: Int,
        val localSequence: Long,
    ) {
        fun toReceipt(generation: Long) = FactReceipt(
            status = "accepted",
            accepted = true,
            committed = false,
            storeGeneration = generation,
            mobileFactId = mobileFactId,
            globalSequence = null,
        )

        fun toItem() = FactPageItem(mobileFactId, payload, partitionId, localSequence, committed = false)
    }

    private class IndexEntry(
        val mobileFactId: String,
        val payload: ByteArray,
        val partitionId: Int,
        val globalSequence: Long,
        val generation: Long,
    ) {
        fun toReceipt(currentGeneration: Long) = FactReceipt(
            status = "committed",
            accepted = true,
            committed = true,
            storeGeneration = currentGeneration,
            mobileFactId = mobileFactId,
            globalSequence = globalSequence,
        )

        fun toItem() = FactPageItem(mobileFactId, payload, partitionId, globalSequence, committed = true)

        fun toJson() = JSONObject()
            .put("mobileFactId", mobileFactId)
            .put("partitionId", partitionId)
            .put("globalSequence", globalSequence)
            .put("generation", generation)

        fun localSequence(): Long {
            val parts = mobileFactId.split(':')
            return parts[2].toLong()
        }
    }

    companion object {
        const val SIDECAR_NAME = "receipt-index-v1.jsonl"
    }
}

private fun hashFromId(mobileFactId: String): String {
    val parts = mobileFactId.split(':')
    return parts[3]
}

File: flutter/ai_app_bridge_flutter/ios/ai_app_bridge_flutter/Sources/AiAppBridgeIOS/Capture/FactStoreReceiptPort.swift
import Foundation

struct FactReceipt: Equatable, Sendable {
    var status: String
    var accepted: Bool
    var committed: Bool
    var storeGeneration: Int64
    var mobileFactId: String?
    var globalSequence: Int64?
}

struct FactCursor: Equatable, Sendable {
    var storeGeneration: Int64
    var afterSequence: Int64 = 0
}

struct FactPageItem: Equatable, Sendable {
    var mobileFactId: String
    var payload: Data
    var partitionId: UInt32
    var globalSequence: Int64
    var committed: Bool
}

struct FactPage: Equatable, Sendable {
    var items: [FactPageItem]
    var storeGeneration: Int64
    var generationMismatch: Bool = false
}

struct DrainResult: Equatable, Sendable {
    var ok: Bool
    var committed: Int
}

struct ThroughWatermark: Equatable, Sendable {
    var throughSequence: Int64
    var storeGeneration: Int64
}

final class FactStoreReceiptPort: @unchecked Sendable {
    static let sidecarName = "receipt-index-v1.jsonl"

    private let store: SegmentedFactStore
    private let sidecar: URL
    private let lock = NSLock()
    private var pending: [String: PendingFact] = [:]
    private var pendingOrder: [String] = []
    private var aliases: [String: String] = [:]
    private var dropped: Set<String> = []
    private var index: [String: IndexEntry] = [:]
    private var indexOrder: [String] = []
    private var generation: Int64 = 1
    private var sequence: Int64 = 0
    private var generationFloor: Int64 = 0
    private var storeHighWater: Int64 = 0

    init(store: SegmentedFactStore, directory: URL) {
        self.store = store
        self.sidecar = directory.appendingPathComponent(Self.sidecarName)
        try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        loadSidecar()
        hydrateFromStore()
        storeHighWater = index.values.map(\.globalSequence).max() ?? 0
    }

    func appendWithReceipt(
        _ payload: Data,
        partitionId: UInt32 = 0,
        durability: String = "async"
    ) -> FactReceipt {
        let enqueue = store.recordForReceipt(payload, partitionId: partitionId, durability: durabilityOf(durability))
        if enqueue != .accepted {
            return receipt(of: enqueue)
        }
        lock.lock()
        defer { lock.unlock() }
        sequence += 1
        let mobileFactId = "mf1:\(generation):\(sequence):\(receiptHashPrefix(payload))"
        pending[mobileFactId] = PendingFact(
            mobileFactId: mobileFactId,
            payload: payload,
            partitionId: partitionId,
            localSequence: sequence
        )
        pendingOrder.append(mobileFactId)
        return FactReceipt(
            status: "accepted",
            accepted: true,
            committed: false,
            storeGeneration: generation,
            mobileFactId: mobileFactId,
            globalSequence: nil
        )
    }

    func commitWait(mobileFactId: String, timeoutMs: Int) -> FactReceipt {
        if !drainWriter(timeoutMs: timeoutMs) {
            lock.lock()
            defer { lock.unlock() }
            return receiptAfterWait(mobileFactId)
        }
        promoteCommitted(markUnmatchedDropped: true)
        lock.lock()
        defer { lock.unlock() }
        return receiptAfterWait(mobileFactId)
    }

    func readPage(cursor: FactCursor, limit: Int) -> FactPage {
        lock.lock()
        defer { lock.unlock() }
        if cursor.storeGeneration != generation {
            return FactPage(items: [], storeGeneration: generation, generationMismatch: true)
        }
        var items: [FactPageItem] = []
        for id in indexOrder {
            guard let entry = index[id] else { continue }
            if entry.globalSequence <= cursor.afterSequence { continue }
            if items.count == limit { break }
            items.append(entry.item())
        }
        for id in pendingOrder {
            if items.count == limit { break }
            guard let held = pending[id] else { continue }
            items.append(held.item())
        }
        return FactPage(items: items, storeGeneration: generation)
    }

    func flushDrain(timeoutMs: Int) -> DrainResult {
        if !awaitFlush(timeoutMs: timeoutMs) {
            return DrainResult(ok: false, committed: 0)
        }
        lock.lock()
        let before = index.count
        lock.unlock()
        promoteCommitted(markUnmatchedDropped: true)
        lock.lock()
        let after = index.count
        lock.unlock()
        return DrainResult(ok: true, committed: after - before)
    }

    func clear() -> Int64 {
        let floor = readAllStoreRecords().map(\.sequence).max().map(Int64.init) ?? 0
        _ = store.takeReceiptOutcomes()
        lock.lock()
        defer { lock.unlock() }
        generation += 1
        sequence = 0
        generationFloor = max(generationFloor, floor)
        storeHighWater = max(storeHighWater, generationFloor)
        pending.removeAll()
        pendingOrder.removeAll()
        aliases.removeAll()
        dropped.removeAll()
        index.removeAll()
        indexOrder.removeAll()
        let row = #"{"op":"clear","generation":\#(generation),"throughSequence":\#(generationFloor)}"# + "\n"
        appendSidecar(row)
        return generation
    }

    func storeGeneration() -> Int64 {
        lock.lock()
        defer { lock.unlock() }
        return generation
    }

    func throughWatermark() -> ThroughWatermark {
        lock.lock()
        defer { lock.unlock() }
        return ThroughWatermark(
            throughSequence: index.values.map(\.globalSequence).max() ?? 0,
            storeGeneration: generation
        )
    }

    private func receiptAfterWait(_ mobileFactId: String) -> FactReceipt {
        if dropped.contains(mobileFactId) {
            return droppedReceipt(mobileFactId, generation: generation)
        }
        if let held = pending[mobileFactId] {
            return held.receipt(generation: generation)
        }
        if let aliased = aliases[mobileFactId], let entry = index[aliased] {
            return entry.receipt(currentGeneration: generation)
        }
        if let entry = index[mobileFactId] {
            return entry.receipt(currentGeneration: generation)
        }
        return droppedReceipt(mobileFactId, generation: generation)
    }

    private func promoteCommitted(markUnmatchedDropped: Bool) {
        let records = readAllStoreRecords()
        let outcomes = store.takeReceiptOutcomes()
        lock.lock()
        defer { lock.unlock() }
        let taken = Set(index.values.map(\.globalSequence))
        let fresh = records
            .map { (sequence: Int64($0.sequence), record: $0) }
            .filter { $0.sequence > generationFloor && !taken.contains($0.sequence) }
            .sorted { $0.sequence < $1.sequence }
        let successCount = outcomes.filter { $0 }.count
        let newlyWritten = successCount == 0 ? [] : Array(fresh.suffix(successCount))
        let preexisting = successCount == 0 ? fresh : Array(fresh.dropLast(successCount))
        var remaining = pending
        var remainingOrder = pendingOrder
        var promoted: [IndexEntry] = []
        for item in preexisting {
            promoted.append(commitStoreRecord(item.record, acceptId: nil))
        }
        var writtenIndex = 0
        var outcomeIndex = 0
        for id in pendingOrder {
            guard remaining[id] != nil else { continue }
            if outcomeIndex >= outcomes.count { break }
            let success = outcomes[outcomeIndex]
            outcomeIndex += 1
            remaining.removeValue(forKey: id)
            remainingOrder.removeAll { $0 == id }
            if !success {
                if markUnmatchedDropped {
                    dropped.insert(id)
                } else if let held = pending[id] {
                    remaining[id] = held
                    remainingOrder.append(id)
                }
                continue
            }
            if writtenIndex >= newlyWritten.count {
                if markUnmatchedDropped { dropped.insert(id) }
                continue
            }
            let record = newlyWritten[writtenIndex].record
            writtenIndex += 1
            promoted.append(commitStoreRecord(record, acceptId: id))
        }
        while writtenIndex < newlyWritten.count {
            promoted.append(commitStoreRecord(newlyWritten[writtenIndex].record, acceptId: nil))
            writtenIndex += 1
        }
        if markUnmatchedDropped {
            for id in remaining.keys {
                dropped.insert(id)
            }
            remaining.removeAll()
            remainingOrder.removeAll()
        }
        pending = remaining
        pendingOrder = remainingOrder
        storeHighWater = max(storeHighWater, promoted.map(\.globalSequence).max() ?? 0)
        for entry in promoted {
            appendSidecar(entry.jsonLine())
        }
    }

    private func commitStoreRecord(_ record: SegmentedFactStoreRecord, acceptId: String?) -> IndexEntry {
        let sequence = Int64(record.sequence)
        let hash = receiptHashPrefix(record.payload)
        let committedId = "mf1:\(generation):\(sequence):\(hash)"
        if let acceptId, acceptId != committedId {
            aliases[acceptId] = committedId
        }
        let entry = IndexEntry(
            mobileFactId: committedId,
            payload: record.payload,
            partitionId: record.partitionId,
            globalSequence: sequence,
            generation: generation
        )
        index[committedId] = entry
        if !indexOrder.contains(committedId) {
            indexOrder.append(committedId)
        }
        return entry
    }

    private func appendSidecar(_ row: String) {
        ensureSidecarLineBoundary()
        guard let data = row.data(using: .utf8) else { return }
        if FileManager.default.fileExists(atPath: sidecar.path) {
            if let handle = try? FileHandle(forWritingTo: sidecar) {
                handle.seekToEndOfFile()
                handle.write(data)
                try? handle.close()
            }
        } else {
            try? data.write(to: sidecar)
        }
    }

    private func ensureSidecarLineBoundary() {
        guard let data = try? Data(contentsOf: sidecar), !data.isEmpty else { return }
        if data.last == UInt8(ascii: "\n") { return }
        guard let handle = try? FileHandle(forWritingTo: sidecar) else { return }
        handle.seekToEndOfFile()
        handle.write(Data("\n".utf8))
        try? handle.close()
    }

    private func readAllStoreRecords() -> [SegmentedFactStoreRecord] {
        var records: [SegmentedFactStoreRecord] = []
        var cursor = SegmentedFactStoreCursor()
        while true {
            guard let page = awaitRead(cursor) else { break }
            if page.isEnd || page.record == nil { break }
            records.append(page.record!)
            cursor = page.cursor
        }
        return records
    }

    private func drainWriter(timeoutMs: Int) -> Bool {
        waitForStore(timeoutMs: timeoutMs) { completion in
            store.status { _ in completion(true) }
        }
    }

    private func awaitFlush(timeoutMs: Int) -> Bool {
        waitForStore(timeoutMs: timeoutMs) { completion in
            store.flush { result in completion(result.isSuccess) }
        }
    }

    private func waitForStore(timeoutMs: Int, start: (@escaping (Bool) -> Void) -> Void) -> Bool {
        let lock = NSLock()
        var done = false
        var ok = false
        start { success in
            lock.lock()
            done = true
            ok = success
            lock.unlock()
        }
        let deadline = Date().addingTimeInterval(Double(timeoutMs) / 1000)
        while Date() < deadline {
            lock.lock()
            let finished = done
            let success = ok
            lock.unlock()
            if finished { return success }
            Thread.sleep(forTimeInterval: 0.01)
        }
        lock.lock()
        let finished = done
        let success = ok
        lock.unlock()
        return finished && success
    }

    private func awaitRead(_ cursor: SegmentedFactStoreCursor) -> SegmentedFactStoreReadResult? {
        var capacity = 64 * 1024
        for _ in 0..<2 {
            guard let page = awaitReadOnce(cursor, bufferCapacity: capacity) else { return nil }
            if page.operation.code != SegmentedFactStoreResultCode.bufferTooSmall { return page }
            if page.requiredCapacity <= capacity { return page }
            capacity = page.requiredCapacity
        }
        return awaitReadOnce(cursor, bufferCapacity: capacity)
    }

    private func awaitReadOnce(
        _ cursor: SegmentedFactStoreCursor,
        bufferCapacity: Int
    ) -> SegmentedFactStoreReadResult? {
        let lock = NSLock()
        var value: SegmentedFactStoreReadResult?
        store.read(cursor: cursor, bufferCapacity: bufferCapacity) { result in
            lock.lock()
            value = result
            lock.unlock()
        }
        let deadline = Date().addingTimeInterval(5)
        while Date() < deadline {
            lock.lock()
            let current = value
            lock.unlock()
            if current != nil { return current }
            Thread.sleep(forTimeInterval: 0.01)
        }
        return nil
    }

    private func loadSidecar() {
        guard let text = try? String(contentsOf: sidecar, encoding: .utf8) else { return }
        for line in text.split(separator: "\n", omittingEmptySubsequences: false) {
            if line.isEmpty { continue }
            guard let data = line.data(using: .utf8),
                  let row = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { continue }
            if row["op"] as? String == "clear" {
                generation = (row["generation"] as? NSNumber)?.int64Value ?? generation
                sequence = 0
                generationFloor = max(generationFloor, (row["throughSequence"] as? NSNumber)?.int64Value ?? 0)
                index.removeAll()
                indexOrder.removeAll()
                continue
            }
            guard let mobileFactId = row["mobileFactId"] as? String else { continue }
            let entry = IndexEntry(
                mobileFactId: mobileFactId,
                payload: Data(),
                partitionId: UInt32((row["partitionId"] as? NSNumber)?.intValue ?? 0),
                globalSequence: (row["globalSequence"] as? NSNumber)?.int64Value ?? 0,
                generation: (row["generation"] as? NSNumber)?.int64Value ?? 1
            )
                if entry.generation == generation {
                    index[mobileFactId] = entry
                    if !indexOrder.contains(mobileFactId) {
                        indexOrder.append(mobileFactId)
                    }
                    sequence = max(sequence, entry.localSequence())
                }
        }
    }

    private func hydrateFromStore() {
        lock.lock()
        let empty = index.isEmpty
        lock.unlock()
        if empty { return }
        let records = readAllStoreRecords()
        lock.lock()
        defer { lock.unlock() }
        let bySequence = Dictionary(uniqueKeysWithValues: records.map { (Int64($0.sequence), $0) })
        var hydrated: [String: IndexEntry] = [:]
        for (id, entry) in index {
            guard let record = bySequence[entry.globalSequence] else { continue }
            guard receiptHashPrefix(record.payload) == hashFromId(id) else { continue }
            hydrated[id] = IndexEntry(
                mobileFactId: entry.mobileFactId,
                payload: record.payload,
                partitionId: record.partitionId,
                globalSequence: Int64(record.sequence),
                generation: entry.generation
            )
        }
        index = hydrated
    }

    private func receipt(of enqueue: SegmentedFactRecordEnqueueResult) -> FactReceipt {
        let status: String
        switch enqueue {
        case .queueFull: status = "queue-full"
        case .closed: status = "closed"
        case .disabled: status = "disabled"
        case .payloadTooLarge: status = "payload-too-large"
        case .accepted: status = "accepted"
        }
        lock.lock()
        let current = generation
        lock.unlock()
        return FactReceipt(
            status: status,
            accepted: false,
            committed: false,
            storeGeneration: current,
            mobileFactId: nil,
            globalSequence: nil
        )
    }

    private func droppedReceipt(_ mobileFactId: String, generation: Int64) -> FactReceipt {
        FactReceipt(
            status: "dropped",
            accepted: false,
            committed: false,
            storeGeneration: generation,
            mobileFactId: mobileFactId,
            globalSequence: nil
        )
    }

    private func durabilityOf(_ value: String) -> SegmentedFactStoreDurability {
        if value == "sync" { return .sync }
        if value == "async" { return .memory }
        preconditionFailure(value)
    }
}

private struct PendingFact {
    var mobileFactId: String
    var payload: Data
    var partitionId: UInt32
    var localSequence: Int64

    func receipt(generation: Int64) -> FactReceipt {
        FactReceipt(
            status: "accepted",
            accepted: true,
            committed: false,
            storeGeneration: generation,
            mobileFactId: mobileFactId,
            globalSequence: nil
        )
    }

    func item() -> FactPageItem {
        FactPageItem(
            mobileFactId: mobileFactId,
            payload: payload,
            partitionId: partitionId,
            globalSequence: localSequence,
            committed: false
        )
    }
}

private struct IndexEntry {
    var mobileFactId: String
    var payload: Data
    var partitionId: UInt32
    var globalSequence: Int64
    var generation: Int64

    func receipt(currentGeneration: Int64) -> FactReceipt {
        FactReceipt(
            status: "committed",
            accepted: true,
            committed: true,
            storeGeneration: currentGeneration,
            mobileFactId: mobileFactId,
            globalSequence: globalSequence
        )
    }

    func item() -> FactPageItem {
        FactPageItem(
            mobileFactId: mobileFactId,
            payload: payload,
            partitionId: partitionId,
            globalSequence: globalSequence,
            committed: true
        )
    }

    func jsonLine() -> String {
        #"{"mobileFactId":"\#(mobileFactId)","partitionId":\#(partitionId),"globalSequence":\#(globalSequence),"generation":\#(generation)}"# + "\n"
    }

    func localSequence() -> Int64 {
        let parts = mobileFactId.split(separator: ":")
        guard parts.count > 2, let value = Int64(parts[2]) else {
            preconditionFailure(mobileFactId)
        }
        return value
    }
}

private func receiptHashPrefix(_ bytes: Data) -> String {
    var hash: UInt32 = 0x811c9dc5
    for byte in bytes {
        hash ^= UInt32(byte)
        hash = hash &* 16777619
    }
    return String(format: "%08x", hash)
}

private func hashFromId(_ mobileFactId: String) -> String {
    String(mobileFactId.split(separator: ":")[3])
}
